### PR TITLE
#34 Support Update Statement

### PR DESCRIPTION
**Pull request by OmniaGM**

### CODE REVIEW

File: troy-schema/src/main/scala/troy/schema/SchemaEngine.scala [hunk 1]
```diff
@@ -117,10 +130,61 @@ case class SchemaEngineImpl(schema: Schema, context: Option[KeyspaceName]) exten
       )).map(_.flatten)
     }
 
+  private def extractVariableTypes(s: UpdateStatement): Result[Seq[DataType]] =
+    schema.getTable(s.tableName).flatMap { table =>
+      V.merge(Seq(
+        extractVariablesFomSet(table, s.set),
+        extractVariablesFromUpdateParam(table, s.using),
```
Comment (tabdulradi) [inline troy-schema/src/main/scala/troy/schema/SchemaEngine.scala]: Wrong order. Variables in `USING` should appear before variables in `SET`.
Please first write test cases to expose the problem.

File: troy-schema/src/main/scala/troy/schema/SchemaEngine.scala [hunk 2]
```diff
@@ -117,10 +130,61 @@ case class SchemaEngineImpl(schema: Schema, context: Option[KeyspaceName]) exten
...
+        if (columnName1 == columnName2)
+          extractVariablesFromEitherListLiteralOrBindMarker(table, columnName1, literalOrBindMarker)
+        else
+          V.error(Messages.ListLiteralAssignmentSyntaxError)
+      case _ => noVariables
+    }).map(_.flatten)
+
+  private def extractVariablesFromEitherListLiteralOrBindMarker(table: Table, columnName: Identifier, literalOrBindMarker: Either[ListLiteral, BindMarker]): Result[Seq[DataType]] = {
+    literalOrBindMarker match {
+      case Left(listLiteral) => extractVariablesFromListLiteralAssignment(table, columnName, listLiteral)
```
Comment (tabdulradi) [inline troy-schema/src/main/scala/troy/schema/SchemaEngine.scala]: Please add a todo to validate that column is actually a list.

File: troy-schema/src/main/scala/troy/schema/SchemaEngine.scala [hunk 3]
```diff
@@ -117,10 +130,61 @@ case class SchemaEngineImpl(schema: Schema, context: Option[KeyspaceName]) exten
...
+        case DataType.List(t) => V.success(Seq(DataType.List(t)))
+        case other            => V.error(Messages.ListLiteralAssignmentFailure(columnName, other))
+      }
+    }
+  }
+
+  private def extractVariablesFromTermAssigment(table: Table, columnName: Identifier, updateOperator: Update.UpdateOperator, term: Term): Result[Seq[DataType]] =
+    table.getColumn(columnName).map(_.dataType).flatMap {
+      case DataType.List(t)   => extractVariablesFromTerm(term, DataType.List(t))
+      case DataType.Map(k, t) => extractVariablesFromTerm(term, DataType.Map(k, t))
```
Comment (tabdulradi) [inline troy-schema/src/main/scala/troy/schema/SchemaEngine.scala]: What about `DataType.Set`?

**Comment by OmniaGM** [comment]
@tabdulradi done, check the new commits please.

**Comment by tabdulradi** [comment]
Please write a test case for the variable order bug first.
Also, please an Integration test under the troy-macro subproject to ensure Update is fully working.

**Pull request merged**